### PR TITLE
feat: add useMaterial and useContactMaterial

### LINKED ISSUE #64: Enable adding new ContactMaterial to World
Adding multiple ContactMaterials to the World is useful for unique contact equation properties between two specific Materials

```js
const material1 = new CANNON.Material({ name: 'ElasticMaterial1', friction: 0, restitution: 1 })
const material2 = new CANNON.Material({ name: 'ElasticMaterial2', friction: 0.2, restitution: 0.9 })
const contactMaterial = new CANNON.ContactMaterial(material1, material2, {})
world.addContactMaterial(contactMaterial)
world.addBody(new CANNON.Body({
    // ...props
    material1
}))
world.addBody(new CANNON.Body({
    // ...props
    material2
}))
```

### PR DESCRIPTION
**Pull request by alexandernanberg**
Closes #64

Uses parts of #122

### CODE REVIEW

File: examples/vite.config.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ export default defineConfig({
     alias: {
       // Resolve symlink ourselves
       '@react-three/fiber': resolve('node_modules', '@react-three', 'fiber'),
+      '@react-three/cannon': resolve('../'),
```
Comment (alexandernanberg) [inline examples/vite.config.js:12]: I had to add this for it to correctly link to the local build

**Comment by hazem3500** [comment]
Will this PR be merged soon 😁?

File: examples/vite.config.js (hunk 1 above)
Comment (bjornstar) [reply examples/vite.config.js:12]:
```suggestion
      '@react-three/cannon': resolve('node_modules', '@react-three', 'cannon'),
```

File: examples/src/demos/Friction.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  return (
+    <mesh ref={ref} receiveShadow>
+      <planeBufferGeometry args={[10, 5]} />
+      <meshStandardMaterial color="lightsalmon" />
+    </mesh>
+  )
+}
+
+function PhysicsContent() {
+  const groundMaterial = useMaterial(() => ({ friction: 0.5 }))
```
Comment (bjornstar) [inline examples/src/demos/Friction.tsx:29]: How is `useMaterial` better than passing materialOptions to one of the physics body hooks?

File: src/hooks.ts [hunk 3]
```diff
@@ -839,3 +840,17 @@ export function useRaycastVehicle(
   }, deps)
   return [ref, api]
 }
+
+type MaterialFn = () => MaterialOptions
+
+export function useMaterial(fn?: MaterialFn): Material {
+  const [material] = useState(() => new Material(fn && fn()))
```
Comment (bjornstar) [inline src/hooks.ts:847]: We are not using state to hold onto physics entities in the hooks as they live inside of the worker.

File: src/propsToBody.js [hunk 4]
```diff
@@ -75,7 +75,7 @@ const propsToBody = (uuid, props, type) => {
     ...extra,
     mass: bodyType === 'Static' ? 0 : mass,
     type: bodyType ? Body[bodyType.toUpperCase()] : undefined,
-    material: material ? new Material(material) : undefined,
+    material: material ? (material instanceof Material ? material : new Material(material)) : undefined,
```
Comment (bjornstar) [inline src/propsToBody.js:78]: I'm pretty sure this instanceof check will never pass. Messages passed to the worker are serialized and are only primitive types.

**Comment by bjornstar** [comment]
I appreciate you trying to revive a recently closed PR, but it had some fundamental issues that are still not resolved here. The primary problem is that we never use Cannon on the client side and only use it in the Worker. We synchronize properties between the two and track individual entities using GUIDs.

Implementing useContactMaterial will require tracking GUIDs of the materials.

**Comment by alexandernanberg** [comment]
Unfortunately I won't have to to continue this in the near future, might pick it up again at some point but closing for now

**Comment by Glavin001** [comment]
I have an open PR adding ContactMaterial support and a `Friction` example: https://github.com/pmndrs/use-cannon/pull/316
Would appreciate any code reviews!

**Pull request closed without merging**